### PR TITLE
[Kubernetes provider]Update kubernetes provider README

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
## What does this PR do?

This PR updates the README of Kubernetes Provider. It was created years ago and is out of date.  It includes high level information and configuration options of Kubernetes Provider.

## Why is it important?

The information provided in the README are out of date and can confuse a lot of readers.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

### CODE REVIEW

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 1]
```diff
@@ -1,20 +1,32 @@
...
-Template based autodiscover of Kubernetes resources is only supported in standalone mode as of now.
-It is not part of the Kubernetes Integration yet.
+  
 
-Hints based autodiscover is not supported yet.
+Kubernetes provider can be configured both in **standalone** and **fleet managed** elastic agent.
 
-### Template based autodiscover
+a. In  **standalone** elastic agent, user needs to edit the [elastic-agent-standalone-kubernetes.yaml](https://github.com/elastic/elastic-agent/blob/f994f5bfdf68db27902a4175c3b655b4d611cf7c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L28)
+b. In **fleet managed** elastic agent, kubernetes provider is enabled by default with default values. User needs to follow the steps described in the [documentation](https://www.elastic.co/guide/en/fleet/current/advanced-kubernetes-managed-by-fleet.html) to configure it.
```
Comment (pkoutsovasilis) [inline internal/pkg/composable/providers/kubernetes/README.md:12]: I feel like a blank line is missing from here

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 2]
```diff
@@ -1,20 +1,32 @@
...
+
+Hints based autodiscover is only supported in **standalone** elastic agent(see [doc](https://www.elastic.co/guide/en/fleet/current/hints-annotations-autodiscovery.html)).
+
+  
+
+### Conditions based autodiscover
+
+  
 
 Example:
```
Comment (pkoutsovasilis) [inline internal/pkg/composable/providers/kubernetes/README.md:24]: with the new write up below Example: feels redundant

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
-As an example we will use again redis module.
-In agent.yml(configmap) an extra input block needs to be added.
+
+As an example we will use the redis module.
+
+To automatically identify a Redis Pod and monitor it with the Redis integration, add the following input configuration inside the [Elastic Agent Standalone manifest](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml):
```
Comment (pkoutsovasilis) [inline internal/pkg/composable/providers/kubernetes/README.md:28]: qq: for the Conditions-based autodiscover to work correctly is any kubernetes provider required to be enabled? If yes please provide the necessary settings of the provider in this example

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 4]
```diff
@@ -43,28 +55,50 @@ In agent.yml(configmap) an extra input block needs to be added.
...
+kind: Pod
+metadata:
+  name: redis 
+  annotations:
+    co.elastic.hints/package: redis
+    co.elastic.hints/data_streams: info
+    co.elastic.hints/info.period: 5m
+```
+
+In order to enable hints based autodiscover, user needs to uncomment the following lines in the [Elastic Agent Standalone manifest](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml)
```
Comment (pkoutsovasilis) [inline internal/pkg/composable/providers/kubernetes/README.md:97]: point to the lines in the k8s yaml (https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L31) or even better, if you can, capture which agent providers configuration is needed and then point to the lines

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 5]
```diff
@@ -43,28 +55,50 @@ In agent.yml(configmap) an extra input block needs to be added.
...
+2. Add the [volumeMount](https://github.com/elastic/elastic-agent/blob/c01636e7383a9b2af9a588e0fcf1a4cae7d0d65c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L783-L785) and [volume](https://github.com/elastic/elastic-agent/blob/c01636e7383a9b2af9a588e0fcf1a4cae7d0d65c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L824-L826).
+
+The init container will start before the elastic agent pod and will donwload all the templates of packages [supported](https://github.com/elastic/elastic-agent/tree/main/deploy/kubernetes/elastic-agent-standalone/templates.d). 
+Elastic agent will then collect from the pods running(through the watchers mechanism) all the hints annotations and will try to match them with the correct package.
+In the redis example, it will use the [redis template](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone/templates.d/redis.yml) and substitute the template variables with the values specified in the annotations. Default values will be used for variables not provided.
+A new input will be then created for redis and reds data will start being collected.
```
Comment (pkoutsovasilis) [inline internal/pkg/composable/providers/kubernetes/README.md:104]:
```suggestion
A new input will be then created for redis and redis-specific data will start being collected.
```

File: internal/pkg/composable/providers/kubernetes/README.md (hunk 3 above)
Comment (blakerouse) [reply internal/pkg/composable/providers/kubernetes/README.md:28]: The default settings of the kubernetes provider should be enough for conditions to work.

File: internal/pkg/composable/providers/kubernetes/README.md (hunk 1 above)
Comment (blakerouse) [reply internal/pkg/composable/providers/kubernetes/README.md:12]: Does that cause it to render different? I don't think so.

**Review by blakerouse — APPROVED** [review]
A very nice improvement, thank you.

File: internal/pkg/composable/providers/kubernetes/README.md (hunk 1 above)
Comment (pkoutsovasilis) [reply internal/pkg/composable/providers/kubernetes/README.md:12]: it actually does, now **a.** and **b.** are in the same paragraph which makes it kind hard to distinguish the two. However if you add an empty line between **a.** and **b.** they appear properly separated 🙂

Comment (blakerouse) [reply internal/pkg/composable/providers/kubernetes/README.md:12]: Then absolutely fix before merge.

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 6]
```diff
@@ -1,20 +1,32 @@
...
 
-Template based autodiscover of Kubernetes resources is only supported in standalone mode as of now.
-It is not part of the Kubernetes Integration yet.
+  
 
-Hints based autodiscover is not supported yet.
+Kubernetes provider can be configured both in **standalone** and **fleet managed** elastic agent.
 
-### Template based autodiscover
+a. In  **standalone** elastic agent, user needs to edit the [elastic-agent-standalone-kubernetes.yaml](https://github.com/elastic/elastic-agent/blob/f994f5bfdf68db27902a4175c3b655b4d611cf7c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L28)
```
Comment (swiatekm) [inline internal/pkg/composable/providers/kubernetes/README.md:11]: Isn't this provider enabled by default in general? I believe this is the case for all providers.

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 7]
```diff
@@ -43,28 +55,50 @@ In agent.yml(configmap) an extra input block needs to be added.
...
-10. It then updates the [emitter](https://github.com/elastic/beats/blob/3c77c9a92a2e90b85f525293cb4c2cfc5bc996b1/x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go#L111) with them.
-11. The emitter controller will update the ast that is then used by the agent to generate the final inputs and spawn new programs to deploy the changes([code](https://github.com/elastic/beats/blob/3c77c9a92a2e90b85f525293cb4c2cfc5bc996b1/x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go#L151)).
+
+The Kubernetes provider watches for Kubernetes resources and generates mappings from them (similar to events in beats provider). The mappings include those variables([list of variables](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html#_provider_for_pod_resources)) for each k8s resource with unique value for each one of them.
+
+Agent composable controller which controls all the providers receives these mappings and tries to match them with the input blogs of the configurations.
```
Comment (swiatekm) [inline internal/pkg/composable/providers/kubernetes/README.md:70]:
```suggestion
Agent composable controller which controls all the providers receives these mappings and tries to match them with the input blocks of the configurations.
```

Also, maybe add a link to the controller module here?

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 8]
```diff
@@ -43,28 +55,50 @@ In agent.yml(configmap) an extra input block needs to be added.
...
+
+This means that for every mapping that the condition matches (`kubernetes.pod.labels.app == redis`), a new input will be created in which the condition will be removed(not needed anymore) and the `kubernetes.pod.ip` variable will be substituted from the value in the same mapping.
+
+The updated complete inputs block will be then forwarded to agent to spawn/update metricbeat and filebeat instances.
+
+
+### Hints based autodiscover
+
+Standalone elastic agent supports autodiscover based on hints collected from the [Kubernetes Provider](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html). The hints mechanism looks for hints in kubernetes pod annotations that have the prefix `co.elastic.hints`. As soon as the container starts, elastic agent checks it for hints and launches the proper configuration for the container. Hints tell elastic agent how to monitor the container by using the proper integration.
```
Comment (swiatekm) [inline internal/pkg/composable/providers/kubernetes/README.md:79]:
```suggestion
Standalone elastic agent supports autodiscover based on hints collected from the [Kubernetes Provider](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html). The hints mechanism looks for hints in kubernetes pod annotations that have the prefix `co.elastic.hints`. As soon as the Pod is ready, elastic agent checks it for hints and launches the proper configuration for the container. Hints tell elastic agent how to monitor the container by using the proper integration.
```

File: internal/pkg/composable/providers/kubernetes/README.md (hunk 6 above)
Comment (MichaelKatsoulis) [reply internal/pkg/composable/providers/kubernetes/README.md:11]: You mean if the user removes at all the `kubernetes.provider` block? I don't know then. Maybe I should state it differently, as user does not need to edit the configuration. It is there(in the manifest) with default values.

File: internal/pkg/composable/providers/kubernetes/README.md (hunk 1 above)
Comment (MichaelKatsoulis) [reply internal/pkg/composable/providers/kubernetes/README.md:13]: A blank line is needed indeed

**Comment by MichaelKatsoulis** [comment]
@blakerouse , @pkoutsovasilis and @swiatekm  thank you a lot for your detailed review. I have updated the PR with all your comments. Can you take a final look?

**Review by pkoutsovasilis — APPROVED** [review]
thanks for taking care the comments @MichaelKatsoulis , LGTM

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 9]
```diff
@@ -1,20 +1,30 @@
...
 
-Hints based autodiscover is not supported yet.
+Kubernetes provider can be configured both in **standalone** and **fleet managed** elastic agent.
 
-### Template based autodiscover
+a. In  **standalone** elastic agent, the kubernetes provider is configured inside the [elastic-agent-standalone-kubernetes.yaml](https://github.com/elastic/elastic-agent/blob/f994f5bfdf68db27902a4175c3b655b4d611cf7c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L28). User can update the parameters based on their needs.
+
+b. In **fleet managed** elastic agent, kubernetes provider is enabled by default with default values. User needs to follow the steps described in the [documentation](https://www.elastic.co/guide/en/fleet/current/advanced-kubernetes-managed-by-fleet.html) to configure it.
+
+Condition based autodiscover is supported both in **fleet managed** elastic agent and in **standalone** elastic agent(see [doc](https://www.elastic.co/guide/en/fleet/current/conditions-based-autodiscover.html)).
```
Comment (andrzej-stencel) [inline internal/pkg/composable/providers/kubernetes/README.md:15]:
```suggestion
Condition based autodiscover is supported both in **fleet managed** elastic agent and in **standalone** elastic agent (see [doc](https://www.elastic.co/guide/en/fleet/current/conditions-based-autodiscover.html)).
```

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 10]
```diff
@@ -43,28 +53,54 @@ In agent.yml(configmap) an extra input block needs to be added.
...
-10. It then updates the [emitter](https://github.com/elastic/beats/blob/3c77c9a92a2e90b85f525293cb4c2cfc5bc996b1/x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go#L111) with them.
-11. The emitter controller will update the ast that is then used by the agent to generate the final inputs and spawn new programs to deploy the changes([code](https://github.com/elastic/beats/blob/3c77c9a92a2e90b85f525293cb4c2cfc5bc996b1/x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go#L151)).
+
+The Kubernetes provider watches for Kubernetes resources and generates mappings from them (similar to events in beats provider). The mappings include those variables([list of variables](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html#_provider_for_pod_resources)) for each k8s resource with unique value for each one of them.
```
Comment (andrzej-stencel) [inline internal/pkg/composable/providers/kubernetes/README.md:66]:
```suggestion
The Kubernetes provider watches for Kubernetes resources and generates mappings from them (similar to events in beats provider). The mappings include those variables ([list of variables](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html#_provider_for_pod_resources)) for each k8s resource with unique value for each one of them.
```

**Review by andrzej-stencel — APPROVED** [review]
I feel like "Elastic Agent" should be capitalized, but I'm not a tech writing expert.

**Review by gizas — APPROVED** [review]

File: internal/pkg/composable/providers/kubernetes/README.md (hunk 6 above)
Comment (swiatekm) [reply internal/pkg/composable/providers/kubernetes/README.md]: I believe it's enabled by default if no configuration is provided, but your current edit is fine.

File: internal/pkg/composable/providers/kubernetes/README.md [hunk 11]
```diff
@@ -43,28 +53,54 @@ In agent.yml(configmap) an extra input block needs to be added.
...
+kind: Pod
+metadata:
+  name: redis 
+  annotations:
+    co.elastic.hints/package: redis
+    co.elastic.hints/data_streams: info
+    co.elastic.hints/info.period: 5m
+```
+
+In order to enable hints based autodiscover, user needs to uncomment the following lines in the [Elastic Agent Standalone manifest](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml)
```
Comment (swiatekm) [inline internal/pkg/composable/providers/kubernetes/README.md:95]: @pkoutsovasilis we should make a note to update this with Helm Chart specific instructions after it's released.

**Review by swiatekm — APPROVED** [review]

**Comment by pierrehilbert** [comment]
@MichaelKatsoulis as this is only updating our README, shouldn't we also backport this to 8.17?

**Pull request merged**